src/iosource/Manager.cc

```diff
@@ -59,124 +109,411 @@ IOSource* Manager::FindSoonest(double* ts)
 			break;
 			}
 
-	// Ideally, we would always call select on the fds to see which
-	// are ready, and return the soonest. Unfortunately, that'd mean
-	// one select-call per packet, which we can't afford in high-volume
-	// environments.  Thus, we call select only every SELECT_FREQUENCY
-	// call (or if all sources report that they are dry).
+	// If there aren't any sources and exit_only_after_terminate is false, just
+	// return an empty set of sources. We want the main loop to end.
+	if ( Size() == 0 && ( ! BifConst::exit_only_after_terminate || terminating ) )
+		return ready;
 
-	++call_count;
+	double timeout = -1;
+	IOSource* timeout_src = nullptr;
+	timer_expired = false;
 
-	IOSource* soonest_src = 0;
-	double soonest_ts = 1e20;
-	double soonest_local_network_time = 1e20;
-	bool all_idle = true;
+	PktSrc* pkt_src = pkt_srcs.empty() ? nullptr : pkt_srcs.front();
 
-	// Find soonest source of those which tell us they have something to
-	// process.
-	for ( SourceList::iterator i = sources.begin(); i != sources.end(); ++i )
+	// If we have an active source of data then find the next timeout value from
+	// those sources. Start by getting the next timeout from the timer manager
+	// since it's not an IOSource.
+	// TODO: how do we handle multiple timer managers here?
+	if ( pkt_src || broker_mgr->Active() )
 		{
-		if ( ! (*i)->src->IsIdle() )
+		timeout = timer_mgr->GetNextTimeout();
+
+		for ( auto src : sources )
 			{
-			all_idle = false;
-			double local_network_time = 0;
-			double ts = (*i)->src->NextTimestamp(&local_network_time);
-			if ( ts >= 0 && ts < soonest_ts )
+			if ( src->src->IsOpen() )
 				{
-				soonest_ts = ts;
-				soonest_src = (*i)->src;
-				soonest_local_network_time =
-					local_network_time ?
-						local_network_time : ts;
+				double next = src->src->GetNextTimeout();
+				if ( timeout == -1 || ( next >= 0.0 && next < timeout ) )
+					{
+					timeout = next;
+					timeout_src = src->src;
+					}
 				}
 			}
 		}
 
-	// If we found one and aren't going to select this time,
-	// return it.
-	int maxx = 0;
+	// If the active packet source isn't live or the packet source is ready (i.e.
+	// we're in pseudo-realtime mode and it's time for a packet), always insert
+	// the packet source into the list of ready sources.
+	if ( pkt_src && pkt_src->IsOpen() && ! pkt_src->IsLive() && pkt_src->GetNextTimeout() == 0 )
+		ready.insert(pkt_src);
+
+	// Call the appropriate poll method for what's available on the operating system.
+	timer_expired = Poll(ready, timeout, timeout_src);
 
-	if ( soonest_src && (call_count % SELECT_FREQUENCY) != 0 )
-		goto finished;
+	return ready;
+	}
+
+#if defined(HAVE_EPOLL_H)
```

Comment:
This is low priority, but I think it's worth considering factoring the logic for the 3 cases out of the mgr into a separate system-specific helper class.

Also, any thoughts on a stragegy for testing these 3 cases? Seems something that could easily break on some system without anybody noticing for a while. If we factored this out (per above) maybe this is a good use case for the new unit testing framework?

Reply:
I mentioned that in a comment shortly after I opened the PR:

> Another note is that I did look at breaking up iosource::Manager into separate classes based on the queue type and what #defines exist, but it ends up being a bunch of extra checks for the defines. It might be more understandable like that, but that's one caveat.

I agree that that would a much cleaner design. It'd still require at least one place to check for what's supported in a factory method, maybe. But it would be better that what's here now. That said, it's low-hanging fruit and I'll revisit it after everything else is taken care of.